src/chain/shared_state.rs

```diff
@@ -172,3 +167,61 @@ pub enum SectionChange {
     Splitting,
     Merging,
 }
+
+/// Vec-like container that is guaranteed to contain at least one element.
+#[derive(PartialEq, Eq)]
+pub struct NonEmptyList<T> {
+    head: Vec<T>,
+    tail: T,
+}
+
+impl<T> NonEmptyList<T> {
+    pub fn new(first: T) -> Self {
+        Self {
+            head: Vec::new(),
+            tail: first,
+        }
+    }
+
+    pub fn push(&mut self, item: T) {
+        self.head.push(mem::replace(&mut self.tail, item))
+    }
+
+    pub fn len(&self) -> usize {
+        self.head.len() + 1
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = &T> + DoubleEndedIterator {
+        self.head.iter().chain(iter::once(&self.tail))
+    }
+
+    pub fn last(&self) -> &T {
+        &self.tail
+    }
+
+    // Retains only elements matching the predicate. If no element matches, the last one is
+    // retained anyway, to maintain the non-empty invariant.
+    pub fn retain<F>(&mut self, mut f: F)
```

Comment:
I'm really not so sure about that NonEmptyList.
We now have this wide contract for retain that somehow keep the last element.

It seems strange to me as a level of abstraction.
It seems either we should have a more specific OurInfo object with private member(s) and the invariant held by the interface and providing roughly this interface without retain, but instead provide a `keep_version_above` that would not fully empty the vector. (You could then have a simpler function than `retain` because you know to keep the `tail` always). 

Or make the member of SharedState private and hold invariant there.